src/gradualizer.erl

```diff
@@ -83,23 +80,500 @@ type_check_dir(Dir, Opts) ->
+      name       := Func,
+      arity      := Arity
+    }) ->
+    io_lib:format("Call to undefined function ~p:~p/~p on line ~p~n",
+              [Module, Func, Arity, LINE]);
+format_error(#{error_type := {undef, record},
+      line     := LINE,
+      module   := Module,
+      name     := RecName
+    }) ->
+    io_lib:format("Undefined record ~p:~p on line ~p~n",
+              [Module, RecName, LINE]);
+format_error(#{error_type := {undef, record},
+      line     := LINE,
+      name     := RecName
+    }) ->
+    io_lib:format("Undefined record ~p on line ~p~n",
+              [RecName, LINE]);
+format_error(#{error_type := {undef, Type},
+      line       := LINE,
+      module     := Module,
+      name       := Name,
+      arity      := Arity
+    })
+  when Type =:= user_type; Type =:= remote_type ->
+    io_lib:format("Undefined ~p ~p:~p/~p on line ~p~n",
+              [Type, Module, Name, Arity, LINE]);
+format_error(#{error_type := {undef, user_type},
+      line       := LINE,
+      name       := Name,
+      arity      := Arity
+    }) ->
+    io_lib:format("Undefined user type ~p/~p on line ~p~n",
+              [Name, Arity, LINE]);
+format_error(#{error_type := unexported_remote_type,
+      line       := LINE,
+      module     := Module,
+      name       := Name,
+      arity      := Arity
+      }) ->
+    io_lib:format("The type ~s:~s/~p on line ~p is not exported~n",
+              [erl_pp:expr(Module), erl_pp:expr(Name), Arity, LINE]);
+format_error(#{error_type := var,
+      line          := LINE,
+      name          := Var,
+      actual_type   := VarTy,
+      expected_type := Ty
+      }) ->
+    io_lib:format("The variable ~p on line ~p has type ~s "
+              "but is expected to have type ~s~n",
+              [Var, LINE, typelib:pp_type(VarTy), typelib:pp_type(Ty)]);
+format_error(#{error_type := char,
+      line          := LINE,
+      expression    := Char,
+      expected_type := Ty
+      }) ->
+    io_lib:format("The character ~p on line ~p does not have type ~s~n"
+             ,[erl_pp:expr(Char), LINE, typelib:pp_type(Ty)]);
+format_error(#{error_type := atom,
+      line          := LINE,
+      expression    := A,
+      expected_type := Ty
+      }) ->
+    io_lib:format("The atom ~p on line ~p does not have type ~s~n",
+              [A, LINE, typelib:pp_type(Ty)]);
+format_error(#{error_type := string,
+      line          := LINE,
+      expression    := String,
+      expected_type := Ty
+      }) ->
+    io_lib:format("The string ~p on line ~p does not have type ~s~n",
+              [String, LINE, typelib:pp_type(Ty)]);
+format_error(#{error_type := int,
+      line          := LINE,
+      expression    := I,
+      expected_type := Ty
+      }) ->
+    io_lib:format("The integer ~p on line ~p does not have type ~s~n",
+              [I, LINE, typelib:pp_type(Ty)]);
+format_error(#{error_type := float,
+      line          := LINE,
+      expression    := F,
```

Comment:
I think we need to handle that `expression` and most of the fields are optionals. If we want some fields to be mandatory depending on the value of the type_error field, it should be reflected in the types. Otherwise, throwing type errors will be very error-prone.

But (rather than going back to a union type just to be able to express this dependency) I suggest we make format_error capable of printing all the information in the type_error, regardless of whether it makes sense or not. Only the order of fields needs to be defined somewhere. Then we just need one pretty-printer per field.

Example:

```Erlang
format_error(Error = #{error_type = T}) ->
    [format_error_type(T),
     lists:map(fun (F) -> F(Error) end,
               [format_expression/1,        %% "" when expression is not present
                format_expected_type/1,
                format_actual_type/1, ...])].

Reply:
We could of course start specializing the types, e.g.
```Erlang
type type_error() ::
    #{error_type := string,
      line           := erl_anno:anno(),
      expected_type  := type(),
      expression     := string() % I guess...
} | % the rest but excluding string
```
I still find the map very valuable since it lets us add fields and reading code using it is a lot easier than tuples.

Reply:
I thought about that, but I think a union of maps can be too complex. Especially it might not be obvious that some key is mandatory in all maps in the union.